### ISSUE TITLE
update selection every frame in "density_from_universe"

### ISSUE THREAD

**Issue by Balasubra** [MDAnalysis/mdanalysis#584]
Friends,
This issue is in reference to the discussion in MDAnalysis user forum (refer to the following link)
https://groups.google.com/forum/#!topic/mdnalysis-discussion/V6yicCbu70Q

I used density_from_universe function to calculate the water density within a cylindrical layer around a protein channel. By default, the function tracks only those water molecules within a cylindrical layer, identified in the first frame. Hence other waters that enter or exit the cylin. layer during the dynamics are omitted.

Suggested Fix: I slightly modified the function by adding an extra user option called, `sel_update=None` (line 426 of `density.py`)
```python
def density_from_Universe(universe, delta=1.0, atomselection='name OH2',
                          metadata=None, padding=2.0, cutoff=0, soluteselection=None,
                          use_kdtree=True, sel_update=None,  **kwargs):
```
Correspondingly, few lines were added while looping over the trajectory
```python
for ts in u.trajectory:
    print("Histograming %6d atoms in frame %5d/%d  [%5.1f%%]\r" % \
               (len(coord), ts.frame, u.trajectory.n_frames, 100.0 * ts.frame / u.trajectory.n_frames),)
   #Added by Bala
    if sel_update==True:
            group = u.select_atoms(atomselection)
            coord = group.coordinates()
    else: coord = current_coordinates()
```
If the user choose `sel_update=True`, then the water selection is updated for each frame. I have tested the modified code, when `sel_update= None` (default), the print statement reports same number of atoms and while with `sel_update=True`, i see varying number of atoms in each frame.

However when `sel_update=True`, the code is slow as expected. I hope the developers may have a better way to implement the functionality.

**Comment by richardjgowers** [comment]
Hi Bala

This sounds like a sensible way to solve this problem.  If you can put together a pull request it would be easier to see if there was a more efficient way of calculating the selection.

https://github.com/MDAnalysis/mdanalysis/wiki/ContributingCode

Selections are something which will be getting much faster soon (with #363), so this fix might be less slow in the future too.

**Comment by orbeckst** [comment]
@Balasubra can you please submit a pull reques ("PR") if you're interested in getting this new feature into MDAnalysis? We are pretty much doing all development and code review through this mechanism, even if it looks as if it is only a single line. With the project getting more and more contributions we need to adhere to some common standards to remain efficient while at the same time upholding code quality.

Thanks!

**Comment by Balasubra** [comment]
I truly apologize that i missed to do that under the stress of a review report. 
Just a naive question, how do i do that ?. I went to pull request, tried to create a new pull request. It it showed me a set of options like : issue-363, h-pages, feature-quantifiedcode etc.

**Comment by Balasubra** [comment]
I have now created a pull request.

**Comment by dotsdl** [comment]
@Balasubra you will need to clone the repository to your own Github account, make changes to a branch there, with name `issue-584`, and then make the pull request from that branch to our `develop` branch. See our [contributing code](https://github.com/MDAnalysis/mdanalysis/wiki/ContributingCode) guide for more detail on how to do this.

**Comment by Balasubra** [comment]
Now i am able to figure out how to do that. I hope now i have done it correctly.

**Comment by orbeckst** [comment]
@Balasubra , have you got a new PR ready? 
[Contributing code](/MDAnalysis/mdanalysis/wiki/ContributingCode) should tell you how. Note that you *must*
* add you changes to the [develop branch](/MDAnalysis/mdanalysis/wiki/DevelopmentBranch)
* [create a PR](/MDAnalysis/mdanalysis/pulls) against the MDAnalysis/mdanalysis *develop* branch or use the *Pull Request button* on your fork — see https://help.github.com/articles/using-pull-requests/ for more details; make sure to *compare against the develop branch*.
* Add a comment to the PR that references this issue #584 so that the PR and this issue get linked.

Please ask if you need more help.

**Comment by Balasubra** [comment]
Hi, 
Yes i cloned the repository and created a new pull request a week back, 
https://github.com/Balasubra/mdanalysis/tree/Balasubra-Issue584
https://github.com/Balasubra/mdanalysis/pull/1

I dnt get what else should i do ?. I see an option called: "This branch has no conflicts with the base branch". Should i do merge pull request ?

**Comment by orbeckst** [comment]
You submitted a PR against your own repository and not against MDAnalysis/mdanalysis so the PR never showed up here and therefore nothing happened.

Do the following:

1. Open a new PR at  https://github.com/MDAnalysis/mdanalysis/pulls

2. Use the **compare across forks** option:

   - *base fork*: **MDAnalysis/mdanalysis**
   - *base*: **develop**
   - *head fork*: **Balasubra/mdanalysis**
   - *head*: **Balasubra-issue584**

3. You should now be able to create a PR that only contains a few lines of code changes. Edit title and comments as necessary. In particular, reference issue #584 in the comment field *and* in the title.

4. Once the PR has been submitted, people will look at the changes and possibly ask questions or ask for changes or additions (such as tests). Please monitor the PR and this issue and respond.

Perhaps the following link will work as a shortcut for you and just do steps 1--3 in one go: https://github.com/MDAnalysis/mdanalysis/compare/develop...Balasubra:Balasubra-Issue584?expand=1#

**Comment by Balasubra** [comment]
Yes i have done it via the second route, now i see the pull request.

**Comment by richardjgowers** [comment]
Closed in #614

**Comment by orbeckst** [comment]
sorry @richardjgowers – forgot to manually close. Thanks for taking care of it!
